bigrig/config.py

```diff
@@ -287,54 +238,44 @@ def __eq__(self, other: object) -> bool:
                 str(self_pkg) == str(other_pkg)
                 for self_pkg, other_pkg in zip_longest(self.packages, other.packages)
             )
-            and self.entry == other.entry
+            and self.config == other.config
         )
 
 
 class Settings:
-    root: RootConfig
+    _wrapped: t.Optional[dict] = None
+    _config_dir: t.Optional[str] = None
 
-    def __getattribute__(self, name: str) -> t.Any:
-        try:
-            return super().__getattribute__(name)
-        except AttributeError as exc:
-            if name == "root":
-                raise ImportError(
-                    f"Application is improperly configured. Before accessing 'settings.{name}'"
-                    f" '{__name__}.settings.configure()' needs to called"
-                ) from exc
-            raise
+    @property
+    def configured(self) -> bool:
+        return self._wrapped is not None
 
-    def __eq__(self, other: object) -> bool:
-        if not isinstance(other, Settings):
-            raise TypeError(
-                f"'==' not supported between instances of '{self.__class__.__name__}' and"
-                f" '{other.__class__.__name__}'"
+    def __getattr__(self, name: str) -> t.Any:
+        if not self.configured:
+            raise SettingsNotConfigured(
+                f"Application is not configured. Call '{__name__}.settings.configure() before accessing settings"
             )
-        return self.root == other.root
+        return self._wrapped[name]
 
     def __repr__(self) -> str:
-        try:
-            return f"{self.__class__.__name__}(root='{self.root}')"
-        except ImportError:
+        if self._config_dir is not None:
+            return f"{self.__class__.__name__}({self._config_dir})"
+        elif self.configured:
+            return f"{self.__class__.__name__}"
+        else:
             return f"{self.__class__.__name__}(<Unconfigured>)"
 
-    def configure(self) -> None:
-        try:
-            # Trigger attribute checks
-            self.root
-        except ImportError:
-            # `root` object is missing, configure it
-            super().__setattr__("root", RootConfig.get_instance())
-        else:
-            # `root` object is present, we're calling `configure` 1+ times
-            raise RuntimeError(
-                f"Settings already have been configured once. A duplicate call to"
-                f" '{__name__}.settings.configure()' exists somewhere in the code path"
+    def configure(self, config_dir: str = None) -> None:
+        config_dir = config_dir or os.environ.get("BIGRIG_CONFIG_PATH")
+        if not config_dir:
+            raise ValueError(
+                f"Missing path to bigrig configuration, point BIGRIG_CONFIG_PATH to bigrig"
+                f"configuration file"
             )
 
-
-settings = Settings()
+        root = RootConfig.from_dir(config_dir)
+        self._wrapped = {**root.all_settings()}
```

Comment:
I'm not entirely convinced in the advantages of doing things this way. The main drawback I see and why I did it the other way is that we loose all typing we had and mypy just goes bonkers.